cloudprovider: aws: return true on existence check for stopped instances

sjenning opened the pull request:
xref https://bugzilla.redhat.com/show_bug.cgi?id=1559271
xref https://github.com/openshift/origin/issues/19899

background https://github.com/kubernetes/kubernetes/pull/45986#issuecomment-386332055

Basically our customers are hitting this issue where the Node resource is deleted when the AWS instances stop (not terminate).  If the instances restart, the Nodes lose any labeling/taints.

Openstack cloudprovider already made this change https://github.com/kubernetes/kubernetes/pull/59931

fixes https://github.com/kubernetes/kubernetes/issues/45118 for AWS

**Reviewer note**: valid AWS instance states are `pending | running | shutting-down | terminated | stopping | stopped`.  There might be a case for returning `false` for instances in `pending` and/or `terminated` state.  Discuss!

`InstanceID()` changes from https://github.com/kubernetes/kubernetes/pull/45986 credit @rrati 

@derekwaynecarr @smarterclayton @liggitt @justinsb @jsafrane @countspongebob

sjenning commented:
/release-note-none
/sig-aws

smarterclayton commented:
To be clear, what is the behavior of the other cloud providers (vmware, GCP, etc)?  Are they like AWS or like OpenStack now?

sjenning commented:
https://github.com/kubernetes/kubernetes/pull/51409 implemented `InstanceExistsByProviderID()`  for most of the cloudproviders 

The following DO NOT filter on state:
Azure, CloudStack, GCP, vshpere

The following DO filter on state:
Openstack, AWS

https://github.com/kubernetes/kubernetes/pull/59931 doesn't get the job done for Openstack due to https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/openstack/openstack_instances.go#L125-L128

sjenning commented:
The external AWS cloud provider simply vendors its code from kube atm, so the change to

sjenning commented:
Also note that vsphere DOES filter with the built-in but not with the external.  What a crazy place this is...

https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/vsphere/vsphere.go#L755-L764

derekwaynecarr commented:
I agree w/ this change, and would say that the usage of taints and labels continues to grow and indirect loss of such attributes on a node have a number of unintended consequences for operators.  I think the current ux provided by the AWS cloud provider is confusing end-users and should be changed.

See my prior comment:
https://github.com/kubernetes/kubernetes/pull/45986#issuecomment-386332055

neolit123 commented:
/sig cloud-provider
@kubernetes/sig-aws-misc

ingvagabund commented:
/test pull-kubernetes-e2e-kops-aws

liggitt commented:
Also see the discussion at https://github.com/kubernetes/kubernetes/issues/46442#issuecomment-408483372

This was brought up in sig-aws last week, and is on the agenda for sig-cloud-provider next week

The summary of where the various cloud providers are at currently is a good data point.

liggitt commented:
> Reviewer note: valid AWS instance states are pending | running | shutting-down | terminated | stopping | stopped. There might be a case for returning false for instances in pending and/or terminated state. Discuss!

based on https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html, a 'pending' state can occur while starting up a stopped instance, so it should still return true.

I could see returning false for instances in terminated state.

/cc @d-nishi

derekwaynecarr commented:
per https://github.com/kubernetes/kubernetes/pull/67254 this pr seems like the appropriate follow-on.

derekwaynecarr commented:
@jsafrane can you review and approve? see the reference to prereq approved pr on interface documentation change.

gnufied commented:
@sjenning AWS and Openstack are somewhat unique in this regard IIRC. Only these cloud providers remove the node when a node is shutdown.  

In general writing disruptive tests that simulate this behaviour and make sure nothing breaks has been harder because kops currently deploys everything in a ASG and a shutdown node is automatically terminated and removed from ASG.  I am in general okay with this change but we must keep in mind that, there are almost no e2e tests that capture this change.

gnufied commented:
Hmm looks like behavior of Openstack was fixed sometime back but I do remember it removing a deleted node originally.  GCE also keeps shutdown nodes in the list and does not remove them.

GCE fortunately has many disruptive tests, AWS has zero.

d-nishi commented:
@gnufied @liggitt - agree with the suggested approach here as well. 
we are beginning work on fixing testing at AWS. happy to use this scenario to start with. @gnufied can you help identify the disruptive tests in GCE that we should look at?

gnufied commented:
Storage has -  https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/pd.go#L307 and there are probably more. 

a while back - we wrote some disruptive tests for AWS -  https://github.com/gnufied/kubernetes/commit/3df33c7c8285f8dd608bdd97a722eb1a32f5836b . Specifically https://github.com/gnufied/kubernetes/commit/3df33c7c8285f8dd608bdd97a722eb1a32f5836b#diff-cf6d6091f38a1751fd35b8257efc60a3 but there is no way to run them in current test-infra setup, so we ended up deleting those.

sjenning committed:
cloudprovider: aws: return true on existence check for stopped instances

sjenning commented:
I have updated the commit to exclude instances in `terminated` state

liggitt commented:
/retest

gnufied commented:
The regression we noticed today where volumes were not being detached from AWS is possibly because of - https://github.com/kubernetes/kubernetes/issues/65392 . Before this patch, a stopped node caused deletion of all pods on it and now the pods on stopped nodes stick around in "unknown" but "Running" phase.

gnufied commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -4325,7 +4325,8 @@ func (c *Cloud) findInstanceByNodeName(nodeName types.NodeName) (*ec2.Instance,
 		newEc2Filter("private-dns-name", privateDNSName),
-		newEc2Filter("instance-state-name", "running"),
+		// exclude instances in "terminated" state
+		newEc2Filter("instance-state-name", "pending", "running", "shutting-down", "stopping", "stopped"),
```
minor nit - can we make a constant out of the states we consider "alive" ?

sjenning commented:
Rebased and fixed nit.

Please note that https://github.com/kubernetes/kubernetes/pull/59930 merged about 24 hours ago and removes the state check from `InstanceExistsByProviderID()`.  I'm adding it back to consider instances in `terminated` state as dead and filter them out.

fyi @zetaab

zetaab commented:
@sjenning filtering out terminated instances is fine. However, you could use constants (those are defined in aws ec2 lib)

sjenning commented:
@zetaab good call.  updated.

zetaab reviewed (approved):
/approve

gnufied commented:
/lgtm

gnufied commented:
cc @childsb @derekwaynecarr please add it to 1.12 milestone.

derekwaynecarr commented:
impacts sig node.

/sig node
/milestone v1.12
/priority critical-urgent
